feat: Add nullable vector support in import utility layer

marcelo-cjl opened the pull request:
related: #45993 

Add nullable vector support in import utility layer
    
Key changes:

ImportV2 util:
- Add nullable vector types (FloatVector, Float16Vector, BFloat16Vector,
  BinaryVector, SparseFloatVector, Int8Vector) to AppendNullableDefaultFieldsData()
- Add tests for nullable vector field data appending

CSV/JSON/Numpy readers:
- Add nullPercent parameter to test data generation for better null coverage
- Mark vector fields as nullable in test schemas
- Add test cases for nullable vector field parsing
- Refactor tests to use loop-based approach with 0%, 50%, 100% null percentages

Parquet field reader:
- Add ReadNullableBinaryData() for nullable BinaryVector/Float16Vector/BFloat16Vector
- Add ReadNullableFloatVectorData() for nullable FloatVector
- Add ReadNullableSparseFloatVectorData() for nullable SparseFloatVector
- Add ReadNullableInt8VectorData() for nullable Int8Vector
- Add ReadNullableStructData() for generic nullable struct data
- Update Next() to use nullable read methods when field is nullable
- Add null data validation for non-nullable fields

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
- Core invariant: import must preserve per-row alignment and validity for every field — nullable vector fields are expected to be encoded with per-row validity masks and all readers/writers must emit arrays aligned to original input rows (null entries represented explicitly).
- New feature & scope: adds end-to-end nullable-vector support in the import utility layer — AppendNullableDefaultFieldsData in internal/datanode/importv2/util.go now appends nil placeholders for nullable vectors (FloatVector, Float16Vector, BFloat16Vector, BinaryVector, SparseFloatVector, Int8Vector); parquet reader (internal/util/importutilv2/parquet/field_reader.go) adds ReadNullableBinaryData, ReadNullableFloatVectorData, ReadNullableSparseFloatVectorData, ReadNullableInt8VectorData, ReadNullableStructData and routes nullable branches to these helpers; CSV/JSON/Numpy readers and test utilities updated to generate and validate 0/50/100% null scenarios and mark vector fields as nullable in test schemas.
- Logic removed / simplified: eliminates ad-hoc "parameter-invalid" rejections for nullable vectors inside FieldReader.Next by centralizing nullable handling into ReadNullable* helpers and shared validators (getArrayDataNullable, checkNullableVectorAlignWithDim/checkNullableVectorAligned), simplifying control flow and removing scattered special-case checks.
- No data loss / no regression (concrete code paths): nulls are preserved end-to-end — AppendNullableDefaultFieldsData explicitly inserts nil entries per null row (datanode import append path); ReadNullable*Data helpers return both data and []bool validity masks so callers in field_reader.go and downstream readers receive exact per-row validity; testutil.BuildSparseVectorData was extended to accept validData so sparse vectors are materialized only for valid rows while null rows are represented as missing. These concrete paths ensure null rows are represented rather than dropped, preventing data loss or behavioral regression.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

marcelo-cjl commented:
/ci-rerun-e2e-default

foxspy commented on internal/datanode/importv2/util_test.go:
```diff
@@ -599,8 +641,21 @@ func Test_AppendNullableDefaultFieldsData(t *testing.T) {
+						tt.dataType == schemapb.DataType_BFloat16Vector ||
+						tt.dataType == schemapb.DataType_BinaryVector ||
+						tt.dataType == schemapb.DataType_SparseFloatVector ||
+						tt.dataType == schemapb.DataType_Int8Vector
```
The GetRow test shouldn’t be skipped.

marcelo-cjl replied:
done

marcelo-cjl committed:
Add nullable vector support for data import

This commit extends the import pipeline (parquet, json, csv, numpy) to handle
nullable vectors with sparse storage strategy.

Parquet field reader changes:
- Add ReadNullableBinaryData() for nullable BinaryVector/Float16Vector/BFloat16Vector
- Add ReadNullableFloatVectorData() for nullable FloatVector
- Add ReadNullableInt8VectorData() for nullable Int8Vector
- Add ReadNullableSparseFloatVectorData() for nullable SparseFloatVector
- Add ReadNullableStructData() for nullable struct types
- Update Next() to call nullable readers when field.GetNullable() is true

Datanode import util changes:
- Update AppendNullableDefaultFieldsData() to handle all vector types
  by appending nil for nullable vector fields

Storage insert_data changes:
- Update RowNum() for all vector types to return len(ValidData) when
  Nullable is true, ensuring logical row count is returned

Test util changes:
- Add nullable vector field support in CreateInsertData()
- Update test cases across csv, json, numpy, parquet readers

Signed-off-by: marcelo-cjl <marcelo.chen@zilliz.com>

marcelo-cjl commented on internal/util/importutilv2/parquet/field_reader.go:
```diff
@@ -1045,6 +1193,122 @@ func ReadSparseFloatVectorData(pcr *FieldReader, count int64) (any, error) {
+		}
+	}
+	return &storage.SparseFloatVectorFieldData{
+		SparseFloatArray: schemapb.SparseFloatArray{
```
this is mainly to maintain the consistency of function interface

bigsheeper commented:
/lgtm
lgtm for the import part

marcelo-cjl commented:
feature: #45993

marcelo-cjl commented:
/ci-rerun-ut-cpp

marcelo-cjl commented:
/ci-rerun-ut-cpp

marcelo-cjl commented:
/ci-rerun-e2e-default

marcelo-cjl commented:
/ci-rerun-ut-go

bigsheeper commented:
/lgtm

czs007 reviewed (approved):
/lgtm
